src/pilot_simulation.jl

```diff
@@ -1,21 +1,145 @@
 using PyCall
 
+using Interpolations: interpolate, gradient, Gridded, Linear
+const ∇ = gradient
+
+using Oceananigans
+using Oceananigans.Utils
+
 # Needed to import local modules
 # See: https://github.com/JuliaPy/PyCall.jl/issues/48
 py"""
 import sys
 sys.path.insert(0, ".")
 """
 
+#####
+##### Load large-scale (base state) solution from SOSE
+#####
+
 sose = pyimport("sose_data")
 
 ds2 = sose.open_sose_2d_datasets("/home/alir/cnhlab004/bsose_i122/")
 ds3 = sose.open_sose_3d_datasets("/home/alir/cnhlab004/bsose_i122/")
 
-t = sose.get_times(ds2)
+date_times = sose.get_times(ds2)
+
+lat, lon, days = 190, -55, 10
+
+arch = CPU()
+FT = Float64
+
+τx = sose.get_scalar_time_series(ds2, "oceTAUX", lat, lon, days) |> Array{FT}
+τy = sose.get_scalar_time_series(ds2, "oceTAUY", lat, lon, days) |> Array{FT}
+Qθ = sose.get_scalar_time_series(ds2, "oceQnet", lat, lon, days) |> Array{FT}
+Qs = sose.get_scalar_time_series(ds2, "SFLUX",   lat, lon, days) |> Array{FT}
+
+U = sose.get_profile_time_series(ds3, "UVEL",  lat, lon, days) |> Array{FT}
+V = sose.get_profile_time_series(ds3, "VVEL",  lat, lon, days) |> Array{FT}
+Θ = sose.get_profile_time_series(ds3, "THETA", lat, lon, days) |> Array{FT}
+S = sose.get_profile_time_series(ds3, "SALT",  lat, lon, days) |> Array{FT}
+
+ds2.close()
+ds3.close()
+
+#####
+##### Create linear interpolations for base state solution
+#####
+
+ts = day * (0:days-1) |> collect
+zC = ds3.Z.values
+
+ℑτx = interpolate((ts,), τx, Gridded(Linear()))
+ℑτy = interpolate((ts,), τy, Gridded(Linear()))
+ℑQθ = interpolate((ts,), Qθ, Gridded(Linear()))
+ℑQs = interpolate((ts,), Qs, Gridded(Linear()))
+
+# z coordinate needs to be in increasing order.
+reverse!(zC)
+U = reverse(U, dims=2)
+V = reverse(V, dims=2)
+Θ = reverse(Θ, dims=2)
+S = reverse(S, dims=2)
+
+ℑU = interpolate((ts, zC), U, Gridded(Linear()))
+ℑV = interpolate((ts, zC), V, Gridded(Linear()))
+ℑΘ = interpolate((ts, zC), Θ, Gridded(Linear()))
+ℑS = interpolate((ts, zC), S, Gridded(Linear()))
+
+#####
+##### Set up the grid
+#####
+
+Nx = Ny = Nz = 32
+Lx = Ly = Lz = 100
+topology = (Periodic, Bounded, Bounded)
+grid = RegularCartesianGrid(topology=topology, size=(Nx, Ny, Nz) x=(0, Lx), y=(0, Ly), z=(-Lz, 0))
+
+#####
+##### Set up forcing forcings to
+#####   1. include mean flow interactions in the momentum equation, and
+#####   2. weakly relax tracer fields to the base state.
+#####
+
+# Fu′ = - w′∂z(U) - U∂x(u′) - V∂y(u′)
+# FIXME? Do we need to use the flux form operator ∇·(Uũ′) instead of ũ′·∇U ?
+@inline Fu′(i, j, k, grid, t, ũ′, c′, p) =
```

Comment:
the Flux form would seem to be more consistent?

Reply:
Yes. I guess you end up with three terms, ∂x(Uu′) + ∂y(Uv′) + ∂z(Uw′), instead of just one, w′∂z(U), and then we need to pass the linear interpolation object into an operator (should be doable) so I didn't do it yet.

But I agree it would be more consistent.

Reply:
those terms are already included.

Reply:
If your base state is only a function of (z,t), e.g. U = U(z,t) and V = V(z,t) with W = 0 then u′∂x(U) = 0 and v′∂y(U) = 0, right?

So then ũ′·∇U = w′∂z(U) while ∇·(Uũ′) = ∂x(Uu′) + ∂y(Uv′) + ∂z(Uw′).

Reply:
Hmm, sorry I got confused, you are right. There are *also* terms describing the mean lateral advection of the fluctuation, but these are not at stake here.

I think we should use flux form for all terms (both mean lateral advection of the fluctuation, and the advection of the mean by the fluctuation, which we are discussing here). We probably need to do that to get a correct momentum budget.

I will rewrite the notes.